# Support vector machine-based software reuse prediction

## Objective: To implement SVM from scratch and also compared it with using sklearn's SVM

Source of SVM: https://dzone.com/articles/classification-from-scratch-svm-78

In machine learning, support-vector machines (SVMs, also support-vector networks) are supervised learning models with associated learning algorithms that analyze data used for classification and regression analysis. SVM presents one of the most robust prediction methods, based on the statistical learning framework. Given a set of training examples, each marked as belonging to one or the other of two categories, an SVM training algorithm builds a model that assigns new examples to one category or the other, making it a non-probabilistic binary linear classifier (although methods such as Platt scaling exist to use SVM in a probabilistic classification setting). An SVM model is a representation of the examples as points in space, mapped so that the examples of the separate categories are divided by a clear gap that is as wide as possible. New examples are then mapped into that same space and predicted to belong to a category based on the side of the gap on which they fall.

In addition to performing linear classification, SVMs can efficiently perform a non-linear classification using what is called the kernel trick, implicitly mapping their inputs into high-dimensional feature spaces.


### 1. For all ti in training set:
 ti.w + b <= -1   if yi = -1 

 ti.w + b >= +1 if yi = +1 

or

yi(ti.w+b) >= 1

### 2. for all support vectors (i.e., data points that defines margin)
  ti.w+b = -1    where ti is -ve support vector and yi is -1

  ti.w+b = +1    where ti is +ve support vector and yi is +1

### 3. For decision Boundary i.e., yi(ti.w+b)=0 where ti lies within decision boundary
### 4. The goal is to maximize width (W) or to minimize |w|

W = ((X+ - X-).w)/|w|

### 5. After obtaining the tuned w and b we have

x.w+b = 1 is line passing through +ve support vectors

x.w+b = -1 is line passing through -ve support vectors

x.w+b = 0 is decision boundary

### 6. As you know it is not possible that the support vector lines always pass through support vectors

### 7. Thus, it is a convex optimization issue and will lead to a global minimum

### 8. This is Linear SVM i.e., kernel is linear

#Dataset: Reuse/predicting successful reuse

# Attribute Information:
1.  Project ID {A,B,D,E,F,G,H,I,J,K,L,M,N,O,P,Q,R,S,T,U,V,W,X,Y}
2.  Software Staff {L,M,S}
3.  Overall Staff {L,X,M,S}
4.  Type of Software Production {product-family,isolated}
5.  Software and Product {product,alone,process,NA}
6.  SP maturity {high,middle,low}
7.  Application Domain {TLC,SE-Tools,Bank,Engine_Controller,FMS,ATC,TS,Space Manufacturing,Measurement,Finance,Book-Keeping}
8.  Type of Software {Technical,Business,Embedded-RT,Non-Embedded-RT}
9.  Size of Baseline {L,M,S,not_available}
10. Development Approach {OO,proc,not_available}
11. Staff Experience {high,middle,low,not_available}
12. Top Management Commitment {yes,no}
13. Key Reuse Roles Introduced {yes,no,NA}
14. Reuse Processes Introduced {yes,no,NA}
15. Non-Reuse Processes Modified {yes,no,NA}
16. Repository {yes,NA}
17. Human Factors {yes,no}
18. Reuse Approach {tight,loose,NA}
19. Work Products {D+C,C,R+D+C,NA}
20. Domain Analysis {yes,no,NA}
21. Origin {ex-novo,as-is,reeng,NA}
22. Independent Team {yes,no,NA}
23. When Assests Developed {before,justintime,NA}
24. Qualification {yes,no,NA}
25. Configuration Management {yes,no,NA}
26. Rewards Policy {no,yes}
27. Assests {51_to_100,21_to_50,100+,1_to_20,NA}

# Target classes 
Success or Failure {success,failure}

# Source: http://promise.site.uottawa.ca/SERepository/datasets/reuse.arff

# Tasks:
1. Initially, load arff dataset
2. Apply pre-processing techniques
3. Divide data into training and testing sets.
4. Build SVM model from scratch
5. Test your own SVM model
6. Obtain precision and recall
7. Implement sklearn's model on processed data
8. Compare your SVM model with sklearn's model

## Task 1: Implement linear SVM from scratch  
# Algorithm of Linear SVM
1.  Initialize with random big value of w say(w0,w0) we will decrease it later
2.  Set step size as w0*0.1
3.  A minimum value of b, may increase it during the process

        i.  b will range from (-b0 < b < +b0, step = step*b_multiple)

        ii. It is also computational extensive. Therefore, define b0 wisely
4.  Check for points ti in dataset:

        i.  Check all transformation of w like (w0,w0), (-w0,w0), (w0,-w0), (-w0,-w0)

        ii. if not yi(ti.w+b)>=1 for all points then break

        iii.  Else evaluate |w| and put it in dictionary as key and (w,b) as values
5.  If w<=0 then current step is completed and move to step 6

        Else minimize w as (w0-step,w0-step) and move to step 3
6.  While step not becomes w0*0.001 

        i.  step = step*0.1

        ii. move to step 3

7.  Select (w,b) that contain minimum |w| form the dictionary

## Task 2: Implement sklearn's SVM

## Task 3: Compare your SVM with sklearn's SVM with concluding remarks

# Helping links:

https://pythonprogramming.net/svm-in-python-machine-learning-tutorial/

https://medium.com/deep-math-machine-learning-ai/chapter-3-1-svm-from-scratch-in-python-86f93f853dc

https://stackabuse.com/implementing-svm-and-kernel-svm-with-pythons-scikit-learn/

http://ecomunsing.com/build-your-own-support-vector-machine




## Task 1: Implement linear SVM from scratch 

In [83]:
# Load the libraries
from scipy.io import arff
import pandas as pd
from sklearn import preprocessing
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
from matplotlib import style
import numpy as np
from sklearn.metrics import confusion_matrix
from sklearn import svm

In [2]:
!wget http://promise.site.uottawa.ca/SERepository/datasets/reuse.arff

--2020-10-25 13:44:03--  http://promise.site.uottawa.ca/SERepository/datasets/reuse.arff
Resolving promise.site.uottawa.ca (promise.site.uottawa.ca)... 137.122.24.222
Connecting to promise.site.uottawa.ca (promise.site.uottawa.ca)|137.122.24.222|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 14517 (14K) [text/plain]
Saving to: ‘reuse.arff’

reuse.arff          100%[===================>]  14.18K  --.-KB/s    in 0.08s   

2020-10-25 13:44:03 (179 KB/s) - ‘reuse.arff’ saved [14517/14517]



In [15]:
# Load the arff dataset 
# Shuffel the dataset
df = arff.loadarff('reuse.arff')
data = pd.DataFrame(df[0])
data.head()

Project ID Software Staff  ...     # assests Success or Failure
0       b'A'           b'L'  ...  b'51_to_100'         b'success'
1       b'B'           b'L'  ...  b'51_to_100'         b'success'
2       b'D'           b'L'  ...   b'21_to_50'         b'failure'
3       b'E'           b'L'  ...   b'21_to_50'         b'failure'
4       b'F'           b'L'  ...   b'21_to_50'         b'failure'

[5 rows x 28 columns]

In [16]:
# Shuffling
data=data.sample(frac=1).reset_index(drop=True)
data.head()

Project ID Software Staff  ...     # assests Success or Failure
0       b'W'           b'M'  ...    b'1_to_20'         b'success'
1       b'T'           b'S'  ...    b'1_to_20'         b'failure'
2       b'E'           b'L'  ...   b'21_to_50'         b'failure'
3       b'K'           b'M'  ...       b'100+'         b'success'
4       b'H'           b'M'  ...  b'51_to_100'         b'success'

[5 rows x 28 columns]

In [17]:
# Preprocessing
# Encoding categorical variables (if any)
# Feature Scaling
# Filling missing values (if any)
LE = preprocessing.LabelEncoder()
CateList = data.select_dtypes(exclude="int64").columns
print(CateList)

Index(['Project ID', 'Software Staff', 'Overall Staff',
       'Type of Software Production', 'Software and Product', 'SP maturity',
       'Application Domain', 'Type of Software', 'Size of Baseline',
       'Development Approach', 'Staff Experience', 'Top Management Commitment',
       'Key Reuse Roles Introduced', 'Reuse Processes Introduced',
       'Non-Reuse Processes Modified', 'Repository', 'Human Factors',
       'Reuse Approach', 'Work Products', 'Domain Analysis', 'Origin',
       'Independent Team', 'When Assests Developed', 'Qualification',
       'Configuration Management', 'Rewards Policy', '# assests',
       'Success or Failure'],
      dtype='object')


In [18]:
for i in CateList:
    data[i] = LE.fit_transform(data[i])

In [19]:
print(data.shape)
data.head(15)

(24, 28)


Project ID  Software Staff  ...  # assests  Success or Failure
0           21               1  ...          1                   1
1           18               2  ...          1                   0
2            3               0  ...          2                   0
3            9               1  ...          0                   1
4            6               1  ...          3                   1
5           19               2  ...          0                   1
6           12               1  ...          3                   1
7            8               1  ...          3                   0
8            1               0  ...          3                   1
9           14               2  ...          0                   1
10          17               2  ...          0                   1
11          10               1  ...          3                   1
12           5               0  ...          3                   1
13          11               1  ...          0                   1
14          23               1  ...          0                   0

[15 rows x 28 columns]

In [20]:
df = data.iloc[:,:-1]
mm = MinMaxScaler()
df[:]= mm.fit_transform(df[:])

In [21]:
df.head()

Project ID  Software Staff  ...  Rewards Policy  # assests
0    0.913043             0.5  ...             0.0       0.25
1    0.782609             1.0  ...             0.0       0.25
2    0.130435             0.0  ...             1.0       0.50
3    0.391304             0.5  ...             0.0       0.00
4    0.260870             0.5  ...             0.0       0.75

[5 rows x 27 columns]

In [40]:
# Divide the dataset to training and testing set
X = df.values
y = data['Success or Failure'].values
print(X.shape, y.shape)

(24, 27) (24,)


In [41]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.30, random_state = 123)

In [42]:
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

(16, 27) (16,)
(8, 27) (8,)


In [44]:
y_train = y_train.reshape((16,1))
y_test = y_test.reshape((8,1))
print(y_train.shape, y_test.shape)

(16, 1) (8, 1)


In [55]:
# Implement SVM from scratch

def linear_kernel(x1, x2):
    return np.dot(x1, x2)


def polynomial_kernel(x, y, p=3):
    return (1 + np.dot(x, y)) ** p


def gaussian_kernel(x, y, sigma=5.0):
    numerator = np.linalg.norm(x-y)**2
    denominator = 2 * (sigma ** 2)
    return np.exp(-numerator / denominator)


class SVM(object):

    def __init__(self, kernel=linear_kernel, tol=1e-3, C=0.1,
                 max_passes=5, sigma=0.1):

        self.kernel = kernel
        self.tol = tol
        self.C = C
        self.max_passes = max_passes
        self.sigma = sigma
        self.model = dict()

    def __repr__(self):
        return (f"{self.__class__.__name__}("
                f"kernel={self.kernel.__name__}, "
                f"tol={self.tol}, "
                f"C={self.C}, "
                f"max_passes={self.max_passes}, "
                f"sigma={self.sigma}"
                ")")

    def svmTrain(self, X, Y):
        # Data parameters
        m = X.shape[0]

        # Map 0 to -1
        Y = np.where(Y == 0, -1, 1)

        # Variables
        alphas = np.zeros((m, 1), dtype=float)
        b = 0.0
        E = np.zeros((m, 1), dtype=float)
        passes = 0

        # Pre-compute the kernel matrix
        if self.kernel.__name__ == 'linear_kernel':
            print(f'Pre-computing {self.kernel.__name__} matrix')
            K = X @ X.T

        elif self.kernel.__name__ == 'gaussian_kernel':
            print(f'Pre-computing {self.kernel.__name__} matrix')
            X2 = np.sum(np.power(X, 2), axis=1).reshape(-1, 1)
            K = X2 + (X2.T - (2 * (X @ X.T)))
            K = np.power(self.kernel(1, 0, self.sigma), K)

        else:
            # Pre-compute the Kernel Matrix
            # The following can be slow due to lack of vectorization
            print(f'Pre-computing {self.kernel.__name__} matrix')
            K = np.zeros((m, m))

            for i in range(m):
                for j in range(m):
                    x1 = np.transpose(X[i, :])
                    x2 = np.transpose(X[j, :])
                    K[i, j] = self.kernel(x1, x2)
                    K[i, j] = K[j, i]

        print('Training...')
        print('This may take 1 to 2 minutes')

        while passes < self.max_passes:
            num_changed_alphas = 0

            for i in range(m):

                E[i] = b + np.sum(alphas * Y * K[:, i].reshape(-1, 1)) - Y[i]

                if (Y[i] * E[i] < -self.tol and alphas[i] < self.C) or (Y[i] * E[i] > self.tol and alphas[i] > 0):
                    j = np.random.randint(0, m)
                    while j == i:
                        # make sure i is not equal to j
                        j = np.random.randint(0, m)

                    E[j] = b + np.sum(alphas * Y *
                                      K[:, j].reshape(-1, 1)) - Y[j]

                    # Save old alphas
                    alpha_i_old = alphas[i, 0]
                    alpha_j_old = alphas[j, 0]

                    # Compute L and H by (10) or (11)
                    if Y[i] == Y[j]:
                        L = max(0, alphas[j] + alphas[i] - self.C)
                        H = min(self.C, alphas[j] + alphas[i])
                    else:
                        L = max(0, alphas[j] - alphas[i])
                        H = min(self.C, self.C + alphas[j] - alphas[i])
                    if L == H:
                        # continue to next i
                        continue

                    # compute eta by (14)
                    eta = 2 * K[i, j] - K[i, i] - K[j, j]
                    if eta >= 0:
                        # continue to next i
                        continue

                    # compute and clip new value for alpha j using (12) and (15)
                    alphas[j] = alphas[j] - (Y[j] * (E[i] - E[j])) / eta

                    # Clip
                    alphas[j] = min(H, alphas[j])
                    alphas[j] = max(L, alphas[j])

                    # Check if change in alpha is significant
                    if np.abs(alphas[j] - alpha_j_old) < self.tol:
                        # continue to the next i
                        # replace anyway
                        alphas[j] = alpha_j_old
                        continue

                    # Determine value for alpha i using (16)
                    alphas[i] = alphas[i] + Y[i] * \
                        Y[j] * (alpha_j_old - alphas[j])

                    # Compute b1 and b2 using (17) and (18) respectively.
                    b1 = b - E[i] - Y[i] * (alphas[i] - alpha_i_old) * \
                        K[i, j] - Y[j] * (alphas[j] - alpha_j_old) * K[i, j]

                    b2 = b - E[j] - Y[i] * (alphas[i] - alpha_i_old) * \
                        K[i, j] - Y[j] * (alphas[j] - alpha_j_old) * K[j, j]

                    # Compute b by (19).
                    if 0 < alphas[i] < self.C:
                        b = b1
                    elif 0 < alphas[j] < self.C:
                        b = b2
                    else:
                        b = (b1 + b2) / 2
                    num_changed_alphas = num_changed_alphas + 1

            if num_changed_alphas == 0:
                passes = passes + 1
            else:
                passes = 0

            print('.', end='', flush=True)

        print('\n DONE! ')

        # Save the model
        idx = alphas > 0
        self.model['X'] = X[idx.reshape(1, -1)[0], :]
        self.model['y'] = Y[idx.reshape(1, -1)[0]]
        self.model['kernelFunction'] = self.kernel
        self.model['b'] = b
        self.model['alphas'] = alphas[idx.reshape(1, -1)[0]]
        self.model['w'] = np.transpose(np.matmul(np.transpose(alphas * Y), X))
        # return model

    def svmPredict(self, X):
        if X.shape[1] == 1:
            X = np.transpose(X)

        # Dataset
        m = X.shape[0]
        p = np.zeros((m, 1))
        pred = np.zeros((m, 1))

        if self.model['kernelFunction'].__name__ == 'linear_kernel':
            p = X.dot(self.model['w']) + self.model['b']

        elif self.model['kernelFunction'].__name__ == 'gaussian_kernel':
            # Vectorized RBF Kernel
            # This is equivalent to computing the kernel
            # on every pair of examples
            X1 = np.sum(np.power(X, 2), axis=1).reshape(-1, 1)
            X2 = np.transpose(np.sum(np.power(self.model['X'], 2), axis=1))
            K = X1 + (X2.T - (2 * (X @ (self.model['X']).T)))
            K = np.power(self.model['kernelFunction'](1, 0, self.sigma), K)
            K = np.transpose(self.model['y']) * K
            K = np.transpose(self.model['alphas']) * K
            p = np.sum(K, axis=1)

        else:
            for i in range(m):
                prediction = 0
                for j in range(self.model['X'].shape[0]):
                    prediction = prediction + self.model['alphas'][j] \
                        * self.model['y'][j] * \
                        self.model['kernelFunction'](np.transpose(
                            X[i, :]), np.transpose(self.model['X'][j, :]))

                p[i] = prediction + self.model['b']

        # Convert predictions into 0 and 1
        pred[p >= 0] = 1
        return pred

    def predict(self, X):
        if X.shape[1] == 1:
            X = np.transpose(X)

        # Dataset
        m = X.shape[0]
        p = np.zeros((m, 1))
        pred = np.zeros((m, 1))

        if self.model['kernelFunction'].__name__ == 'linear_kernel':
            p = X.dot(self.model['w']) + self.model['b']

        elif self.model['kernelFunction'].__name__ == 'gaussian_kernel':
            # Vectorized RBF Kernel
            # This is equivalent to computing the kernel
            # on every pair of examples
            X1 = np.sum(np.power(X, 2), axis=1).reshape(-1, 1)
            X2 = np.transpose(np.sum(np.power(self.model['X'], 2), axis=1))
            K = X1 + (X2.T - (2 * (X @ (self.model['X']).T)))
            K = np.power(self.model['kernelFunction'](1, 0, self.sigma), K)
            K = np.transpose(self.model['y']) * K
            K = np.transpose(self.model['alphas']) * K
            p = np.sum(K, axis=1)

        else:
            for i in range(m):
                prediction = 0
                for j in range(self.model['X'].shape[0]):
                    prediction = prediction + self.model['alphas'][j] \
                        * self.model['y'][j] * \
                        self.model['kernelFunction'](np.transpose(
                            X[i, :]), np.transpose(self.model['X'][j, :]))

                p[i] = prediction + self.model['b']

        # Convert predictions into 0 and 1
        pred[p >= 0] = 1
        return pred

In [47]:
# Train and test your SVM models
%%time
model = SVM()
model

CPU times: user 11 µs, sys: 0 ns, total: 11 µs
Wall time: 14.3 µs


In [48]:
model.svmTrain(X_train, y_train)

Pre-computing linear_kernel matrix
Training...
This may take 1 to 2 minutes
.............
 DONE! 


In [57]:
y_predicted = model.predict(X_train)

In [66]:
def confusion_matrix(y_actual, y_predicted):
    tp = 0
    tn = 0
    fp = 0
    fn = 0

    for i in range(len(y_actual)):
        if y_actual[i] > 0:
            if y_actual[i] == y_predicted[i]:
                tp = tp + 1
            else:
                fn = fn + 1
        if y_actual[i] < 1:
            if y_actual[i] == y_predicted[i]:
                tn = tn + 1
            else:
                fp = fp + 1

    cm = [[tn, fp], [fn, tp]]
    accuracy = (tp+tn)/(tp+tn+fp+fn)
    sens = tp/(tp+fn)
    prec = tp/(tp+fp)
    f_score = (2*prec*sens)/(prec+sens)
    return cm, accuracy, f_score, prec, sens

In [59]:
# Evaluate training and testing precision and recall
confusion_matrix, accuracy, f_score, precision, recall = confusion_matrix(y_train, y_predicted)

In [60]:
confusion_matrix

[[4, 1], [0, 11]]

In [62]:
print("Accuracy: ",accuracy)
print("Precision: ",precision)
print("Recall: ",recall)
print("F-Score: ",f_score)

Accuracy:  0.9375
Precision:  0.9166666666666666
Recall:  1.0
F-Score:  0.9565217391304348


In [64]:
y_pred = model.predict(X_test)

In [67]:
confusion_matrix, accuracy, f_score, precision, recall = confusion_matrix(y_test, y_pred)

In [68]:
confusion_matrix

[[3, 1], [0, 4]]

In [69]:
print("Accuracy: ",accuracy)
print("Precision: ",precision)
print("Recall: ",recall)
print("F-Score: ",f_score)

Accuracy:  0.875
Precision:  0.8
Recall:  1.0
F-Score:  0.888888888888889


## Task 2: Implement sklearn's SVM


In [71]:
# Use the preprocessed dataset here
df.head()

Project ID  Software Staff  ...  Rewards Policy  # assests
0    0.913043             0.5  ...             0.0       0.25
1    0.782609             1.0  ...             0.0       0.25
2    0.130435             0.0  ...             1.0       0.50
3    0.391304             0.5  ...             0.0       0.00
4    0.260870             0.5  ...             0.0       0.75

[5 rows x 27 columns]

In [72]:
# Divide the dataset to training and testing set
X = df.values
y = data['Success or Failure'].values
print(X.shape, y.shape)

(24, 27) (24,)


In [73]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.30, random_state = 123)

In [74]:
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

(16, 27) (16,)
(8, 27) (8,)


In [75]:
y_train = y_train.reshape((16,1))
y_test = y_test.reshape((8,1))
print(y_train.shape, y_test.shape)

(16, 1) (8, 1)


In [81]:
# Train SVM model using sklearn's SVM
%%time

clf = svm.SVC(kernel='linear') 
clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)

CPU times: user 1.54 ms, sys: 56 µs, total: 1.6 ms
Wall time: 1.38 ms


/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [82]:
# Evaluate training and testing precision and recall
from sklearn import metrics
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))
print("Precision:",metrics.precision_score(y_test, y_pred))
print("Recall:",metrics.recall_score(y_test, y_pred))

Accuracy: 1.0
Precision: 1.0
Recall: 1.0



## Task 3: Compare your SVM with sklearn's SVM with concluding remarks


### SVM from scratch -

Accuracy:  0.875 

Precision:  0.8

Recall:  1.0 

### Sklearn's SVM -

Accuracy: 1.0

Precision: 1.0

Recall: 1.0

### Concluding Remarks -

It is clear that Sklearn's SVM is performing better than the SVM implemented from scratch, we can see both the models are performing quite well in terms of Recall. Also, I believe that the Sklearn's implementation is optimized and doing well in terms of time and space complexity as its execution time is pretty low when compared to my SVM model. Also, Sklearn library's implementation is quite popular. It has been examined, tested, debugged, and updated by the members of the community, so the likelihood of hidden errors is decreased and the approaches to the algorithm have been optimized with a greater accuracy.
